### PR TITLE
Ports Bloodstains And Footprints from EE

### PR DESCRIPTION
**Pull request by Lyndomen**
## About the PR
Puddles now get everywhere, and footprints can be cleaned up. Ported from https://github.com/Simple-Station/Einstein-Engines/pull/1190/

## Why / Balance
Its neat, I want it, and it will make the station more immersive. Also uh, something about janitors weeping. 

## Technical details
Lotta C#, ported from EE. I moved stuff to EE namespace, good luck us. 

## Media
image
image
image

## Requirements
- [X] I have tested all added content and changes.
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
I mangled the C# a bit to put it in a EE namespace. 

**Changelog**
:cl:
- add: Ported footprints and bloodstains from EE. Sorry Janitors.

### CODE REVIEW

File: Content.Shared/EE/Footprint/PuddleFootPrintsComponent.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+namespace Content.Shared.EE.FootPrint;
+
+[RegisterComponent]
+public sealed partial class PuddleFootPrintsComponent : Component
+{
+    [ViewVariables(VVAccess.ReadWrite)]
```
Comment (MilonPL) [inline Content.Shared/EE/Footprint/PuddleFootPrintsComponent.cs]: these are already ReadWrite by default

File: Content.Shared/EE/Footprint/PuddleFootPrintsComponent.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    public float SizeRatio = 0.2f;
+
+    [ViewVariables(VVAccess.ReadWrite)]
```
Comment (MilonPL) [inline Content.Shared/EE/Footprint/PuddleFootPrintsComponent.cs]: these are already ReadWrite by default

File: Resources/Prototypes/Entities/Effects/puddle.yml [hunk 3]
```diff
@@ -113,6 +113,7 @@
   components:
   - type: Clickable
   - type: Slippery
+
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Effects/puddle.yml]: what is this random line break

File: Resources/Prototypes/Entities/Effects/puddle.yml [hunk 4]
```diff
@@ -165,3 +166,42 @@
   - type: Tag
     tags:
     - DNASolutionScannable
+  - type: PuddleFootPrints #DeltaV- Begin updates from EE Blood Puddle
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Effects/puddle.yml]: trailing space

File: Resources/Prototypes/Entities/Effects/puddle.yml [hunk 5]
```diff
@@ -165,3 +166,42 @@
...
+        layer:
+        - SlipLayer
+        hard: false
+  - type: SolutionContainerManager
+    solutions:
+      step: { maxVol: 2 }
+  - type: FootPrint
+  - type: Puddle
+    solution: step
+  - type: Appearance #DeltaV-End updates from EE Blood Puddle
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Effects/puddle.yml]: trailing space

File: Resources/Prototypes/Entities/Mobs/Species/arachnid.yml [hunk 6]
```diff
@@ -115,6 +115,9 @@
         sprite: "Effects/creampie.rsi"
         state: "creampie_arachnid"
         visible: false
+  - type: FootPrints
+    leftBarePrint: "footprint-left-bare-spider" #DeltaV port from EE, blood splatter
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/arachnid.yml]: trailing space

File: Resources/Prototypes/Entities/Mobs/Species/base.yml [hunk 7]
```diff
@@ -216,6 +216,7 @@
     - DoorBumpOpener
     - AnomalyHost
   - type: PotentialPsionic # DeltaV - organic species can all be psionic
+  - type: FootPrints  #DeltaV port from EE, blood splatter
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/base.yml]: trailing space

File: Resources/Prototypes/Entities/Mobs/Species/arachnid.yml [hunk 8]
```diff
@@ -115,6 +115,9 @@
         sprite: "Effects/creampie.rsi"
         state: "creampie_arachnid"
         visible: false
+  - type: FootPrints
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/arachnid.yml]: comment should start on the type declaration

File: Resources/Prototypes/Entities/Mobs/Species/diona.yml [hunk 9]
```diff
@@ -110,6 +110,9 @@
           32:
             sprite: Mobs/Species/Human/displacement.rsi
             state: jumpsuit-female
+  - type: FootPrints
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/diona.yml]: again trailing space + comment on the type

File: Resources/Prototypes/Entities/Mobs/Species/dwarf.yml [hunk 10]
```diff
@@ -64,6 +64,9 @@
           32:
             sprite: Mobs/Species/Human/displacement.rsi
             state: jumpsuit-female
+  - type: FootPrints
+    leftBarePrint: "footprint-left-bare-dwarf" #DeltaV port from EE, blood splatter
+    rightBarePrint: "footprint-right-bare-dwarf" #DeltaV port from EE, blood splatter
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/dwarf.yml]: trailing space + comment on the type

File: Resources/Prototypes/Entities/Mobs/Species/reptilian.yml [hunk 11]
```diff
@@ -77,7 +77,9 @@
           32:
             sprite: Mobs/Species/Human/displacement.rsi
             state: jumpsuit-female
-
+  - type: FootPrints #DeltaV port from EE, blood splatter
+    leftBarePrint: "footprint-left-bare-lizard" #DeltaV port from EE, blood splatter
+    rightBarePrint: "footprint-right-bare-lizard" #DeltaV port from EE, blood splatter
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/reptilian.yml]: only needs comment at line 80 + trailing space

File: Resources/Prototypes/Entities/Mobs/Species/slime.yml [hunk 12]
```diff
@@ -120,6 +120,9 @@
           32:
             sprite: Mobs/Species/Human/displacement.rsi
             state: jumpsuit-female
+  - type: FootPrints #DeltaV port from EE, blood splatter
+    leftBarePrint: "footprint-left-bare-slime" #DeltaV port from EE, blood splatter
+    rightBarePrint: "footprint-right-bare-slime" #DeltaV port from EE, blood splatter
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/slime.yml]: only one comment + space

File: Resources/Prototypes/Entities/Mobs/Species/vox.yml [hunk 13]
```diff
@@ -130,6 +130,9 @@
           32:
             sprite: Mobs/Species/Vox/displacement.rsi
             state: shoes
+  - type: FootPrints #DeltaV port from EE, blood splatter
+    leftBarePrint: "footprint-left-bare-lizard" #DeltaV port from EE, blood splatter
+    rightBarePrint: "footprint-right-bare-lizard" #DeltaV port from EE, blood splatter
```
Comment (MilonPL) [inline Resources/Prototypes/Entities/Mobs/Species/vox.yml]: only one comment + space

**Review by MilonPL — DISMISSED** [review]

File: Resources/Prototypes/Entities/Effects/puddle.yml (hunk 3 above)
Comment (Lyndomen) [reply Resources/Prototypes/Entities/Effects/puddle.yml]: EE code has a slip multiplier here for whatever reason so I put it in to track differences. No longer necessary I'll remove it when I wake up

**Comment by flowerfall-system** [comment]
Yay yay yay let's DO it!!! :D

**Comment by dvir001** [comment]
This might sound like a small issue, but the folders should be ``_EE`` not ``EE``
This has a major effect on all forks that use code from EE and already has it as ``_EE`` folders, making it harder to cherry-pick in and out.

**Review by deltanedas — CHANGES_REQUESTED** [review]
i believe this was already used on goob and ended up ddosing the server with endless entity spam, if this is the same code it needs either timed despawn like goob added or fixing cleaning methods and limiting the amount of them per tile

**Comment by Lyndomen** [comment]
yes it is same code, I will import https://github.com/Goob-Station/Goob-Station/pull/847 tmrw

**Comment by Lyndomen** [comment]
image
footprints despawn in 60 secs now, and requested tidying up has happened.

**Comment by IamVelcroboy** [comment]
> footprints despawn in 60 secs now, 

Not that this whole PR isn't cool and I understand the performance concerns but doesn't this sort of defeat the purpose of footprints being a source of forensics evidence and creating work for janitors?

**Comment by Lyndomen** [comment]
> > footprints despawn in 60 secs now,
> 
> Not that this whole PR isn't cool and I understand the performance concerns but doesn't this sort of defeat the purpose of footprints being a source of forensics evidence and creating work for janitors?

Yeah, it kinda does, but its maybe better than nothing? idk, Im down to change the timer if we wanna set it to 5 mins or something. All depends on the performance.

**Comment by ThataKat** [comment]
I feel like they should persist a lot longer than 60 seconds, for the same reasons already given

**Comment by deltanedas** [comment]
the main issue that led to it needing despawning is that you can have a stupid amount of them in a small area and cleaning didnt work, if they were limited and cleanable it would be fine

**Comment by MilonPL** [comment]
we can revert it if the performance dies :blunt:

**Comment by deltanedas** [comment]
it WILL die, on goob the janitors couldnt clean them no matter what and it ruined the tps

**Comment by Lyndomen** [comment]
> the main issue that led to it needing despawning is that you can have a stupid amount of them in a small area and cleaning didnt work, if they were limited and cleanable it would be fine

cleaning works fwiw now 
385575512-ac540b4d-dd9f-4b69-abab-4cd83a685fc5

**Comment by Aidenkrz** [comment]
> it WILL die, on goob the janitors couldnt clean them no matter what and it ruined the tps

It was client side TPS :trollface: the visualizer should just be modified to only show the 10 top most footprints or whatever per tile
and cleaning should be fixed

**Review by deltanedas — APPROVED** [review]

**Comment by deltanedas** [comment]
5 tps incident incoming

**Comment by MilonPL** [comment]
i cant wait

**Comment by Emily9031** [comment]
It seems this was expected, but last night during a round there was a ton of blood on the evac shuttle, and the new footsteps caused a ton of performance issues from everything being in such a small space. I confirmed with other players that it seemed to be affecting everyone, even people with very good hardware such as myself.

Maybe detectives can be given a device similar to the t-ray scanner that shows footprints nearby that were cleaned or dried up.

TLDR more optimization needs to be done.

**Comment by MilonPL** [comment]
who could've seen this coming
